host/sgx/loadelf.c

```diff
@@ -713,12 +716,12 @@ static oe_result_t _patch(oe_enclave_image_t* image, size_t enclave_size)
             oe_round_up_to_multiple(image->tbss_size, image->tbss_size);
     }
 
-    if (aligned_size > OE_THREAD_LOCAL_SPACE)
-    {
-        OE_TRACE_ERROR(
-            "Thread-local variables exceed available thread-local space.\n");
-        OE_RAISE(OE_FAILURE);
-    }
+    aligned_size = oe_round_up_to_multiple(aligned_size, OE_PAGE_SIZE);
+    _set_uint64_t_symbol_value(
```

Comment:
This depends on the enclave image including symbols, which I don't think we can rely on ( I know we do it in other places here but I don't know that we should). Is there a way to instead compute the value at enclave compile time so that the number can be built into the enclave image itself?

Reply:
I guess the TLS size is only known after linking.
I can think of two ways:
* Put the variable at a known location within a specific section (e.g., .oeinfo or its own section) and patch it there.
* Let the enclave parse its ELF headers during TLS initialization to get the values.

Reply:
I think both of those are decent solutions, although I'm leaning toward the second one. A 3rd option is to just pass in a value during an initialization ecall.

@CodeMonkeyLeet @anakrish @mingweishih do any of you have a preference?

Reply:
Given that `td_from_tcs_offset` is just a derived value from constants and other values we already patched into the image as part of this function, we could just recalculate and cache it on enclave init. The tricky part is that it gets used very early in the `oe_enter` call in assembly, and I'm not sure it's worth the extra complexity in that codepath, so I think the current approach reasonable.

Reply:
> This depends on the enclave image including symbols, which I don't think we can rely on ( I know we do it in other places here but I don't know that we should). Is there a way to instead compute the value at enclave compile time so that the number can be built into the enclave image itself?

@jhand2 Great point.

We are incorrectly relying on regular symbols instead of dynamic symbols.

A loader does not rely on regular symbols to load an application/shared library. Instead it relies on dynamic symbols.
Dynamic symbols cannot be stripped away.

libcurl.so does not have any regular symbols whereas it has dynamic symbols necessarsy to perform loading/linking:
```bash
$objdump -t /usr/lib/x86_64-linux-gnu/libcurl.so

/usr/lib/x86_64-linux-gnu/libcurl.so:     file format elf64-x86-64

SYMBOL TABLE:
no symbols

$ objdump -T /usr/lib/x86_64-linux-gnu/libcurl.so
...
000000000002bff0 g    DF .text	000000000000000a  CURL_OPENSSL_4 curl_global_init
0000000000038820 g    DF .text	00000000000000b8  CURL_OPENSSL_4 curl_multi_strerror
000000000004e490 g    DF .text	00000000000000ef  CURL_OPENSSL_4 curl_pushheader_byname
000000000002ca60 g    DF .text	0000000000000083  CURL_OPENSSL_4 curl_easy_recv
0000000000036310 g    DF .text	0000000000000189  CURL_OPENSSL_4 curl_multi_perform
0000000000024a50 g    DF .text	0000000000000037  CURL_OPENSSL_4 curl_getenv
00000000000367f0 g    DF .text	0000000000000041  CURL_OPENSSL_4 curl_multi_socket
000000000002c460 g    DF .text	0000000000000055  CURL_OPENSSL_4 curl_easy_getinfo
000000000002c000 g    DF .text	0000000000000089  CURL_OPENSSL_4 curl_global_init_mem
000000000027e4c0 g    DO .bss	00000000000000c8  CURL_OPENSSL_4 curl_jmpenv
0000000000024bd0 g    DF .text	000000000000000c  CURL_OPENSSL_4 curl_escape
000000000000fde0 g    DF .text	0000000000000cf9  CURL_OPENSSL_4 curl_formadd
00000000000520e0 g    DF .text	0000000000000084  CURL_OPENSSL_4 curl_mime_encoder
```

When we strip out enclaves by specifying `-s` to the linker or by using `strip`, it no longer has any symbols and therefore loading fails as you pointed out:
```bash
$ strip echo_enc
$ objdump -t echo_enc

echo_enc:     file format elf64-x86-64

SYMBOL TABLE:
no symbols
```

However it still retains the symbols that we've specified as exported (ie needed for loading):
```bash
objdump -T echo_enc

echo_enc:     file format elf64-x86-64

DYNAMIC SYMBOL TABLE:
0000000000061edc g    DF .text	0000000000000000 memset
0000000000123fa0 g    DF .text	0000000000000000 oe_longjmp
0000000000145104 g    DF .text	0000000000000000 oe_egetkey
000000000006da10 g    DF .text	0000000000000000 abort
000000000035f9e0 g    D  .bss	0000000000000000 __bss_start
00000000001204bc g    DF .text	0000000000000000 oe_continue_execution
0000000000365730 g    D  .bss	0000000000000000 _end
000000000012046f g    DF .text	0000000000000000 oe_restore_partial_context
0000000000144b17 g    DF .text	0000000000000000 oe_rdseed
000000000035f9e0 g    D  .data	0000000000000000 _edata
0000000000016890 g    DF .text	0000000000000005 _start
0000000000016895 g    DF .text	0000000000000005 oe_call_link_enclave
0000000000061eb4 g    DF .text	0000000000000000 memmove
0000000000144b0c g    DF .text	0000000000000000 oe_rdrand
000000000012052b g    DF .text	00000000000001be oe_enter
00000000001203fc g    DF .text	0000000000000000 oe_snap_current_context
0000000000061e80 g    DF .text	0000000000000000 memcpy
0000000000127bb8 g    DF .text	0000000000000000 oe_setjmp
00000000001206e9 g    DF .text	000000000000004a oe_exception_dispatcher
```

To be consistent with the way loaders work, we'd need to
- export all the symbols necessary for loading
- Change the loader to use only dynamic symbols (they are meant for the loader)

I have a PR that does the above changes: https://github.com/openenclave/openenclave/pull/3359

I'd recommend that we go the above approach and make our loader consistent with the way loaders work.
That will allow enclaves to be loaded even if symbols are not available.

Of the two solutions proposed by @thomasten, (1) would  likely be easier to implement. (2) may not be feasible since the host-side loader needs to set the value of `_enclave_rva` so that the enclave can securely determine the enclave start address.

It would be preferable to revisit the above solutions if ever we need the enclave not have any dynamic symbols as well (that likely may not ever be possible since libcxx always exports symbols, and we'd like to pursure multi-module loading which needs dynamic symbols).

For the foreseeable future, changing the loader to rely only on dynamic symbols is the right approach.

Reply:
Thanks @anakrish! I rebased to master and exported td_from_tcs_offset.